bindings/python/build_tools/cmake/iree_py_extension.cmake

```diff
@@ -112,5 +108,145 @@ function(iree_py_extension)
       # For example, foo/bar/ library 'bar' would end up as 'foo::bar'.
       add_library(${_PACKAGE_NS} ALIAS ${_NAME})
     endif()
+    # Defer computing transitive dependencies and calling target_link_libraries()
+    # until all libraries have been declared.
+    # Track target and deps, use in iree_complete_binary_link_options() later.
+    set_property(GLOBAL APPEND PROPERTY _IREE_PY_EXTENSION_NAMES "${_NAME}")
+    set_property(TARGET ${_NAME} PROPERTY DIRECT_DEPS ${_RULE_DEPS})
+  endif()
+endfunction()
+
+# Lists all transitive dependencies of DIRECT_DEPS in TRANSITIVE_DEPS.
```

Comment:
That's what I initially had in mind, but AFAIR @stellaraccident mentioned in the discussion on the ML, that the bindings might get split to a separate repository. In that case, we might need the boilerplate code.

Reply:
**TLDR**: Copy-paste is okay for now, but let's see if we can make it easier for separate repositories to benefit from this without needing to fork it.

---

Ah good point (and ugh, I wish `ALWAYSLINK` was a CMake feature and not something we had to build on top like this). This could also relate to #659 then. If we have to copy this (CMake alwayslink handling) ourselves for something as core as language bindings, then others may have to as well for their own projects.

I think there's a middle ground between "you must use `iree_cc_library` and `iree_cc_binary`" and "your build system is entirely your own responsibility, good luck" that we can find here. For example, these methods could be optionally available to use from dependent projects if you specify your targets in a certain way.

This doesn't need to happen this in PR, since this can stay focused on building more of the Python bindings, but I worry about divergence and the cost that `ALWAYSLINK` adds for other projects.

---

Brainstorming a bit... how about something of this form?

* Extract the first stage of `set_property` calls above to `iree_track_deps(NAME, GROUP_NAME, DIRECT_DEPS)`
* Extract the second stage of `target_link_libraries` to `iree_get_link_args(NAME)` that returns a list or string which you can append into your own `target_link_libraries` (or wrapper) that has the right list of deps and combination of flags. OR to `iree_complete_link_options(GROUP_NAME)` that iterates over the list and does the linking itself.

Reply:
**TL;DR**: I would highly appreciate if we can merge this including the copy-paste and could handle refactoring separately. Just added further configuration to enable testing the python bindings with ctest. Hence, I am confident that refactoring can be done without me breaking the `ALWAYSLINK` stuff again...

---

Yeah, I fully agree. I think it's also worth to take a closer look into MLIR's approach, which implement this within the function [whole_archive_link()](https://github.com/llvm/llvm-project/blob/d5a186a60014dc1a8c979c978cb32aba7ecb9102/mlir/cmake/modules/AddMLIR.cmake#L9--L28). This called prior to `target_link_libraries`, e.g. [here](https://github.com/llvm/llvm-project/blob/d5a186a60014dc1a8c979c978cb32aba7ecb9102/mlir/tools/mlir-translate/CMakeLists.txt#L23-L24) for `mlir-translate`. This might be a way we can follow in a similar manner for #659. I think it's further super helpful to add some functionality which allows to trace the deps. So finally, I really like the idea you brainstormed. I think that's actually the way to go.

Reply:
We might have to re-discuss considering https://github.com/llvm/llvm-project/commit/c64770506b89a2376fe13080bc3b72789e6c752d#diff-6635fa1a5b0b10edf6e511585c7c986d

Reply:
Yeah, if we can remove the need for `alwayslink` then the build systems could get simpler in this respect. Beyond the uses in MLIR core, we also use `alwayslink` in IREE's HAL driver modules, as in https://github.com/google/iree/blob/18acd53ee4f4348bfd5f336835a9855dbbce8972/iree/hal/vulkan/BUILD#L425-L438

The `hal::Driver` modules are only for dynamic registration though, applications that know ahead of time which `hal::Driver`(s) they want to use can directly depend on the appropriate targets (as in the Vulkan sample). See https://github.com/google/iree/blob/18acd53ee4f4348bfd5f336835a9855dbbce8972/iree/hal/driver_registry.h#L37-L44

Even if we keep that use of `alwayslink`, it might be a narrow-enough slice of the build tree that we could special case it instead of using this general (and complex) solution in CMake.